godot-core/src/obj/raw.rs

```diff
@@ -0,0 +1,555 @@
+use crate::obj::{dom, GodotClass};
+use crate::obj::{GdMut, GdRef, InstanceId};
+use crate::storage::InstanceStorage;
+use crate::{engine, out};
+
+/// Low-level bindings for object pointers in Godot.
+///
+/// This should not be used directly, you should either use [`Gd<T>`](super::Gd) or [`Option<Gd<T>>`]
+/// depending on whether you need a nullable object pointer or not.
+#[repr(C)]
+pub struct RawGd<T: GodotClass> {
+    pub(super) obj: *mut T,
+    // Must not be changed after initialization.
+    cached_instance_id: Option<InstanceId>,
+}
+
+impl<T: GodotClass> RawGd<T> {
+    /// Create a new object representing a null in Godot.
+    pub(super) fn null() -> Self {
+        Self {
+            obj: ptr::null_mut(),
+            cached_instance_id: None,
+        }
+    }
+
+    /// Initializes this `RawGd<T>` from the object pointer as a **weak ref**, meaning it does not
+    /// initialize/increment the reference counter.
+    ///
+    /// # Safety
+    ///
+    /// `obj` must be a valid object pointer or a null pointer.
+    pub(super) unsafe fn from_obj_sys_weak(obj: sys::GDExtensionObjectPtr) -> Self {
+        let mut instance_id = None;
+        if !obj.is_null() {
+            let id =
+                unsafe { interface_fn!(object_get_instance_id)(obj as sys::GDExtensionObjectPtr) };
+            instance_id = InstanceId::try_from_u64(id);
+        }
+
+        Self {
+            obj: obj as *mut T,
+            cached_instance_id: instance_id,
+        }
+    }
+
+    /// Initializes this `RawGd<T>` from the object pointer as a **strong ref**, meaning it initializes
+    /// /increments the reference counter and keeps the object alive.
+    ///
+    /// This is the default for most initializations from FFI. In cases where reference counter
+    /// should explicitly **not** be updated, [`from_obj_sys_weak()`](Self::from_obj_sys_weak) is available.
+    ///
+    /// # Safety
+    ///
+    /// `obj` must be a valid object pointer or a null pointer.
+    pub(super) unsafe fn from_obj_sys(obj: sys::GDExtensionObjectPtr) -> Self {
+        Self::from_obj_sys_weak(obj).with_inc_refcount()
+    }
+
+    /// Returns `self` but with initialized ref-count.
+    fn with_inc_refcount(self) -> Self {
+        // Note: use init_ref and not inc_ref, since this might be the first reference increment.
+        // Godot expects RefCounted::init_ref to be called instead of RefCounted::reference in that case.
+        // init_ref also doesn't hurt (except 1 possibly unnecessary check).
+        T::Mem::maybe_init_ref(&self);
+        self
+    }
+
+    /// Returns `true` if the object is null.
+    ///
+    /// This does not check if the object is dead, for that use
+    /// [`instance_id_or_none()`](Self::instance_id_or_none).
+    pub(crate) fn is_null(&self) -> bool {
+        self.obj.is_null() || self.cached_instance_id.is_none()
+    }
+
+    pub(crate) fn instance_id(&self) -> Option<InstanceId> {
+        self.cached_instance_id
+    }
```

Comment:
This should be renamed to `instance_id_unchecked`, otherwise the discrepancy from the `Gd` methods is confusing.